sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/Repeater.java

```diff
@@ -0,0 +1,142 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableSet;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+
+/** Repeats a method invocation when it encounters an error. */
+public class Repeater<InputT, OutputT> {
+
+  /**
+   * {@link Set} of {@link UserCodeExecutionException}s that warrant repeating. A public modifier is
+   * applied to communicate to users of this class which {@link UserCodeExecutionException}s
+   * constitute warrant repeat execution.
+   */
+  public static final Set<Class<? extends UserCodeExecutionException>> REPEATABLE_ERROR_TYPES =
+      ImmutableSet.of(
+          UserCodeRemoteSystemException.class,
+          UserCodeTimeoutException.class,
+          UserCodeQuotaException.class);
+
+  /** Instantiates a {@link Repeater}. */
+  public static <InputT, OutputT> Repeater<InputT, OutputT> of(
+      ThrowableFunction<InputT, OutputT> throwableFunction, Sleeper sleeper, Integer limit) {
+    return new Repeater<>(throwableFunction, sleeper, limit);
+  }
+
+  private final ThrowableFunction<InputT, OutputT> throwableFunction;
+
+  private final Sleeper sleeper;
+  private final int limit;
+
+  private Repeater(
+      ThrowableFunction<InputT, OutputT> throwableFunction, Sleeper sleeper, int limit) {
+    this.throwableFunction = throwableFunction;
+    this.sleeper = sleeper;
+    this.limit = limit;
+  }
+
+  /**
+   * Applies the {@link InputT} to the {@link ThrowableFunction}. If the function throws an
+   * exception that {@link #REPEATABLE_ERROR_TYPES} contains, repeats the invocation up to the
+   * limit, otherwise throws the last exception.
+   */
+  public OutputT apply(InputT input) throws UserCodeExecutionException, InterruptedException {
+    @MonotonicNonNull UserCodeExecutionException lastException = null;
+    for (int numAttempts = 0; numAttempts < limit; numAttempts++) {
+      try {
+        return throwableFunction.apply(input);
+      } catch (UserCodeExecutionException e) {
+        if (!REPEATABLE_ERROR_TYPES.contains(e.getClass())) {
+          throw e;
+        }
+        lastException = e;
+        sleeper.sleep();
+      }
+    }
+    throw checkStateNotNull(lastException);
+  }
+
+  /**
+   * A {@link FunctionalInterface} for executing a {@link UserCodeExecutionException} throwable
+   * function.
+   */
+  @FunctionalInterface
+  public interface ThrowableFunction<InputT, OutputT> {
+    /** Returns the result of invoking this function on the given input. */
+    OutputT apply(InputT input) throws UserCodeExecutionException;
+  }
+
+  /** Interfaces implementation details for pausing an execution. */
+  public interface Sleeper {
+
+    /** Pauses the execution. */
+    void sleep() throws InterruptedException;
```

Comment:
Should we use `org.apache.beam.sdk.util.Sleeper`? Or, if the next comment is considered, one could refactor to take a Backoff instance instead of sleeper.